### PR TITLE
feat/add-zk-account-button

### PR DESCRIPTION
**Pull request by dansofter**
## Description

- WIP on account modal
- useModal hook to simplify modal experience across the whole UI

closes: #216 

---

Before we can merge this PR, please make sure that all the following items have been checked off. If any of the checklist items are not applicable, please leave them but write a little note why.

- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work
- [ ] Re-reviewed files changed in the Github PR explorer

## If PR to `staging`
- [ ] Updated relevant documentation in the code
- [ ] Test code changes manually, and describe your procedure in a comment on this PR
- [ ] Mark this PR with the correct milestone

## If PR to `main`
- [ ] Update the version number in `package.json` and `src/config/common.json`
- [ ] Update the minimum required signer version number in `src/config/common.json` if applicable
- [ ] Test following the procedure in `docs/manual_test.md`
- [ ] Update `CHANGELOG.md`
- [ ] Make sure that all PR's to `manta-signer` and `sdk` on which this PR depends are merged

### CODE REVIEW

File: src/components/Account/index.tsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import React, { useEffect, useState } from 'react';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faCheck, faCopy } from '@fortawesome/free-solid-svg-icons';
+
+import { useSend } from 'pages/SendPage/SendContext';
+import { usePrivateWallet } from 'contexts/privateWalletContext';
+import GradientText from 'components/GradientText';
+import AssetType from 'types/AssetType';
+
+const Account = () => {
```
Comment (Kevingislason) [inline src/components/Account/index.tsx:10]: I would name this component `AccountModalDisplay`, `Account` is a bit ambiguous

File: src/components/Account/index.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        <h2 className="text-xl">My Private Address</h2>
+        <p className="w-80 break-all">{privateAddress}</p>
+        {addressCopied ? (
+          <FontAwesomeIcon icon={faCheck} />
+        ) : (
+          <FontAwesomeIcon
+            icon={faCopy}
+            className="cursor-pointer"
+            onMouseDown={copyToClipboard}
+          />
```
Comment (Kevingislason) [inline src/components/Account/index.tsx:72]: This copy-paste icon is used elsewhere, so it would be good to factor it out into a reusable component `CopyAddressIcon`

File: src/components/Account/index.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                      text={balance.assetType.ticker}
+                    />
+                  </div>
+                  <div className="w-52">{balance.publicBalance.toString()}</div>
+                  <div className="w-52">
+                    {balance.privateBalance.toString()}
+                  </div>
+                </div>
+              ))
+            : null}
```
Comment (Kevingislason) [inline src/components/Account/index.tsx:103]:
```suggestion
          {balances?.length &&
              balances.map((balance: any) => (
                <div className="flex gap-4">
                  <div className="flex items-center gap-3 w-52">
                    <img
                      className="w-8 h-8 rounded-full"
                      src={balance.assetType.icon}
                      alt="icon"
                    />
                    <GradientText
                      className="text-2xl font-bold"
                      text={balance.assetType.ticker}
                    />
                  </div>
                  <div className="w-52">{balance.publicBalance.toString()}</div>
                  <div className="w-52">
                    {balance.privateBalance.toString()}
                  </div>
                </div>
              ))
           }
```

File: src/components/Account/index.tsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+          />
+        )}
+      </div>
+      <div className="mt-8">
+        <h2 className="text-lg">Balances</h2>
+        <div className="mt-4">
+          <div className="flex gap-4">
+            <div className="w-52">Asset</div>
+            <div className="w-52">Public</div>
+            <div className="w-52">Private</div>
```
Comment (Kevingislason) [inline src/components/Account/index.tsx:81]: As you work more on the UI, I think we should have two tabs labeled "Public" and "Private". If the `public` tab is selected we can display a public address instead of a public address. I think displaying public balances underneath a private address may confuse users.

File: src/components/Account/index.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      senderAssetTypeOptions.map(async (assetType: AssetType) => {
+        const publicBalance = await fetchPublicBalance(
+          senderPublicAccount?.address,
+          assetType
+        );
+        const privateBalance = await getSpendableBalance(assetType);
+        return {
+          assetType,
+          publicBalance: publicBalance.toString(),
+          privateBalance: privateBalance.toString()
```
Comment (Kevingislason) [inline src/components/Account/index.tsx:34]: You will want to use the `Balances` class whenever creating or handling balances

File: src/components/Account/index.tsx (hunk 3 above)
Comment (Kevingislason) [reply src/components/Account/index.tsx:103]: For readability purposes, you could also factor out everything in this block into a component called `SingleAssetBalanceDisplay`

File: src/hooks/useModal.tsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  const ModalWrapper = useMemo(
+    () =>
+      ({ children }: { children: React.ReactNode }) => {
+        return open ? (
+          <ModalContainer>
+            <ModalBackDrop
+              onClick={() => {
+                if (!closeDisabled) {
+                  hideModal();
```
Comment (Kevingislason) [inline src/hooks/useModal.tsx:90]: nit: I would move the `if (!closeDisabled)` check into the `hideModal` function, stylistically you want as little JS as possible in your JSX

File: src/hooks/useModal.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            ) : (
+              <div className="relative inline-block px-6 py-4 rounded-lg bg-secondary text-white">
+                {children}
+              </div>
+            )}
+          </ModalContainer>
+        ) : null;
+      },
+    [open, hideModal]
+  );
```
Comment (Kevingislason) [inline src/hooks/useModal.tsx:113]: This component is a bit hard to read, it would be good to factor out some of the internals into separate components  (like `CloseModalButton`)

**Review by Kevingislason — COMMENTED** [review]
Nice work so far, please let me know if you have any questions or if I can help

File: src/components/Account/index.tsx [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+import { faCheck, faCopy } from '@fortawesome/free-solid-svg-icons';
+
+import { useSend } from 'pages/SendPage/SendContext';
+import { usePrivateWallet } from 'contexts/privateWalletContext';
+import GradientText from 'components/GradientText';
+import AssetType from 'types/AssetType';
+
+const Account = () => {
+  const { privateAddress, getSpendableBalance } = usePrivateWallet();
+  const { senderAssetTypeOptions, senderPublicAccount, fetchPublicBalance } =
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/Account/index.tsx]: https://polkadot.js.org/docs/api/examples/promise/listen-to-balance-change

File: src/components/AccountModalDisplay/index.tsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+import { usePublicBalancesContext } from 'contexts/publicBalancesContext';
+import { usePrivateWallet } from 'contexts/privateWalletContext';
+import { useExternalAccount } from 'contexts/externalAccountContext';
+import GradientText from 'components/GradientText';
+import AssetType from 'types/AssetType';
+
+const AccountModalDisplay = () => {
+  const { privateAddress, getSpendableBalance } = usePrivateWallet();
+  const { getPublicBalance } = usePublicBalancesContext();
```
Comment (Kevingislason) [inline src/components/AccountModalDisplay/index.tsx:13]: By convention, "context" is  not included in the `use` function, so this should be named `usePublicBalances`

File: src/components/AccountModalDisplay/index.tsx [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+  // useEffect(() => {
+  //   if (
+  //     senderAssetTypeOptions &&
+  //     senderAssetTypeOptions.length > 0 &&
+  //     senderPublicAccount
+  //   ) {
+  //     fetchPublicBalances();
+  //   }
+  // }, [senderAssetTypeOptions, senderPublicAccount]);
```
Comment (Kevingislason) [inline src/components/AccountModalDisplay/index.tsx:50]: Seems safe to delete now

File: src/components/AccountModalDisplay/index.tsx [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        <h2 className="text-xl">My Private Address</h2>
+        <p className="w-80 break-all">{privateAddress}</p>
+        {addressCopied ? (
+          <FontAwesomeIcon icon={faCheck} />
+        ) : (
+          <FontAwesomeIcon
+            icon={faCopy}
+            className="cursor-pointer"
+            onMouseDown={copyToClipboard}
+          />
```
Comment (Kevingislason) [inline src/components/AccountModalDisplay/index.tsx:73]: Since this code is repeated elsewhere on the front end, you should refactor it into a component `CopyIcon`, which takes `copyText` as a parameter, and contains all state / functions related to copying. So the copyToClipboard function in `CopyIcon` would look like this:

```js
  const copyToClipboard = () => {
    navigator.clipboard.writeText(copyText);
    setAddressCopied(true);
  };
```

File: src/components/AccountModalDisplay/index.tsx [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        <h2 className="text-lg">Balances</h2>
+        <div className="mt-4">
+          <div className="flex gap-4">
+            <div className="w-52">Asset</div>
+            <div className="w-52">Public</div>
+            <div className="w-52">Private</div>
+          </div>
+          {assets.map((asset: AssetType) => {
+            const balance = getPublicBalance(asset);
+            console.log({ balance, asset });
```
Comment (Kevingislason) [inline src/components/AccountModalDisplay/index.tsx:86]: By convention it is best to avoid logical code in JSX. You will want to get balances outside of the JSX here. Also, don't need both `balance` and `asset`, only `balance`, because a `Balance` also contains an `AssetType`

File: src/components/AccountModalDisplay/index.tsx [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+                    src={balance.assetType.icon}
+                    alt="icon"
+                  />
+                  <GradientText
+                    className="text-2xl font-bold"
+                    text={balance.assetType.ticker}
+                  />
+                </div>
+                <div className="w-52">{balance.toString()}</div>
+                {/* <div className="w-52">{balance.privateBalance.toString()}</div> */}
```
Comment (Kevingislason) [inline src/components/AccountModalDisplay/index.tsx:101]: safe to delete?

File: src/contexts/publicBalancesContext.tsx [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+          return {
+            ...prev,
+            [assetType.assetId]: new Balance(
+              assetType,
+              new BN(balance.data.free.toString())
+            )
+          };
+        });
+      });
+      return;
```
Comment (Kevingislason) [inline src/contexts/publicBalancesContext.tsx:50]: For readability, you could factor out this part into a function `subscribeNativeTokenBalanceChange`

File: src/contexts/publicBalancesContext.tsx [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+      address,
+      ({ value }: any) => {
+        const balanceString = value.isEmpty ? '0' : value.balance.toString();
+        setPublicBalances((prev) => {
+          return {
+            ...prev,
+            [assetType.assetId]: new Balance(assetType, new BN(balanceString))
+          };
+        });
+      }
```
Comment (Kevingislason) [inline src/contexts/publicBalancesContext.tsx:63]: Similarly, you could factor out this part into a function `subscribeExternalTokenBalanceChange`

**Review by Kevingislason — DISMISSED** [review]

**Review by Kevingislason — COMMENTED** [review]
err sorry didn't mean to approve yet, but anyway, coming along well

File: src/components/Navbar/ZkAddress.tsx [hunk 16]
```diff
@@ -0,0 +1,25 @@
+import React, { useState } from 'react';
```
Comment (Kevingislason) [inline src/components/Navbar/ZkAddress.tsx:1]: I would call this component `ZkAccountButton` to distinguish it from the address itself, which could be a type

Comment (Kevingislason) [inline src/components/Navbar/ZkAddress.tsx:1]: I would call this component ZkAccountButton to distinguish it from the address itself, which could be a type

File: src/components/Sidebar/index.tsx [hunk 17]
```diff
@@ -0,0 +1,66 @@
+// @ts-nocheck
```
Comment (Kevingislason) [inline src/components/Sidebar/index.tsx:1]: Why did you add this back? I don't think this is being uses anywhere

File: src/components/Navbar/ZkAddress.tsx (hunk 16 above)
Comment (Kevingislason) [inline src/components/Navbar/ZkAddress.tsx:1]: I would call this component ZkAccountButton to distinguish itself from the ZkAccount itself, which could be a type

File: src/components/ZkAccoutModal/index.tsx [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    let total = new Decimal(0);
+    const balances = await Promise.all(
+      assets.map(async (assetType: any) => {
+        const privateBalance = await getSpendableBalance(assetType);
+
+        if (privateBalance) {
+          const usdBalanceString = privateBalance.toUsdString(
+            usdPrices[assetType.coingeckoId]
+              ? usdPrices[assetType.coingeckoId]
+              : new Decimal(0)
```
Comment (Kevingislason) [inline src/components/ZkAccoutModal/index.tsx:32]: You can remove these turnery expressions if `getSpendableBalance` just returned 0 by default if some coin doesn't have any balance

File: src/components/ZkAccoutModal/index.tsx [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+          };
+        }
+
+        return {
+          assetType,
+          usdBalanceString: new Balance(assetType, new Decimal(0)).toUsdString(
+            new Decimal(0)
+          ),
+          privateBalance
+        };
```
Comment (Kevingislason) [inline src/components/ZkAccoutModal/index.tsx:54]: I don't think both of these returns are necessary

File: src/components/ZkAccoutModal/index.tsx [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+            </div>
+          </>
+        ) : (
+          <span className="whitespace-nowrap">You have no zkAccount yet.</span>
+        )}
+      </div>
+      {privateAddress ? (
+        <div className="border border-secondary rounded-xl px-6 py-4 mb-5 text-secondary">
+          {balances && balances.length > 0 ? (
+            balances.map((balance: any) => (
```
Comment (Kevingislason) [inline src/components/ZkAccoutModal/index.tsx:97]: I think you can just do balances?.map and get rid of this check --> `balances && balances.length > 0 ?`

File: src/utils/general/decimal.ts [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+
+const decimalToUsdString = (value) => {
+  return `$${value
+    .toDecimalPlaces(2, Decimal.ROUND_DOWN)
+    .toNumber()
+    .toLocaleString(undefined, {
+      maximumFractionDigits: 2,
+      minimumFractionDigits: 2
+    })}`;
+};
```
Comment (Kevingislason) [inline src/utils/general/decimal.ts:12]: I think we need to create a type for USD balances, using decimals like we've been doing is a bit hacky

File: src/contexts/publicBalancesContext.tsx [hunk 22]
```diff
@@ -0,0 +1,109 @@
+// @ts-nocheck
```
Comment (Kevingislason) [inline src/contexts/publicBalancesContext.tsx:1]: I like the idea of a publicBalancesContext, but this isn't being used anywhere. I guess we can keep it and refactor to use it later.

File: src/contexts/usdPricesContext.tsx [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    const assets = AssetType.AllCurrencies(false);
+    const ids = assets.reduce((res, asset, index) => {
+      return `${res}${asset.coingeckoId}${
+        index < assets.length - 1 ? ',' : ''
+      }`;
+    }, '');
+
+    const res = await axios.get(
+      `https://api.coingecko.com/api/v3/simple/price?ids=${ids}&vs_currencies=usd`
+    );
```
Comment (Kevingislason) [inline src/contexts/usdPricesContext.tsx:23]: I think you need some error handling here in case the request fails

File: src/contexts/usdPricesContext.tsx [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+
+    if (res.data) {
+      const prices = {};
+      Object.keys(res.data).map(
+        (id) => (prices[id] = new Decimal(res.data[id]['usd']))
```
Comment (Kevingislason) [inline src/contexts/usdPricesContext.tsx:28]: We should store balances by our own assetIds, not by coingeckos asset ids

File: src/contexts/publicBalancesContext.tsx [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+        });
+      }
+    );
+  };
+
+  const subscribeBalanceChanges = async (address: string) => {
+    if (!api || !address) {
+      return null;
+    }
+    console.log({ address });
```
Comment (Kevingislason) [inline src/contexts/publicBalancesContext.tsx:71]: You can remove logs

File: src/contexts/publicBalancesContext.tsx [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+    await api.isReady;
```
Comment (Kevingislason) [inline src/contexts/publicBalancesContext.tsx:72]: We don't need to await api.isReady anymore

File: src/components/ZkAccoutModal/index.tsx [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+import CopyPasteIcon from 'components/CopyPasteIcon';
+
+const AccountDisplay = () => {
+  const { privateAddress, getSpendableBalance, signerIsConnected, isReady } =
+    usePrivateWallet();
+  const { usdPrices } = useUsdPrices();
+
+  const assets = AssetType.AllCurrencies(true);
+  const [balances, setBalances] = useState<any>([]);
+  const [totalBalance, setTotalBalance] = useState('$0.00');
```
Comment (Kevingislason) [inline src/components/ZkAccoutModal/index.tsx:20]: if this is a string, I would call this 'totalBalanceString'

**Review by Kevingislason — COMMENTED** [review]
Nice work. A few things:

-Contents of components with borders (asset list, "You have no zkAccount yet") should be centered
-The borders in Chloe's design are slightly less rounded and grey
-The dollar value text below "Total Balance" is larger in Chloe's design
-I think we should show the Manta icon next to the zk address display. Chloe's design shows a polkadot address icon, but that isn't compatible with zk addresses.
-Some text isn't visible in light mode
-After a transaction, balances in the modal don't update until I open and close the modal
-Private balances disappear for a short time after I sign an transaction
-I think the modal needs to be a little wider. If the user has over $1,000,000 of some asset, the layout gets quite cramped
-We shouldn't actually display USD values for testnet tokens
-Users should be able to close the ZkAccountModal by clicking on the ZkAccountButton

File: src/utils/general/decimal.ts (hunk 21 above)
Comment (Boyuan-Deng) [reply src/utils/general/decimal.ts:12]: DId not have too much context on this. May want to elaborate a little bit.

File: src/components/ZkAccoutModal/index.tsx [hunk 28]
```diff
@@ -0,0 +1,192 @@
...
+                        </div>
+                      </div>
+                    </div>
+                    <div className="text-black dark:text-white">
+                      {usdPrices[balance.assetType.baseTicker]
+                        ? balance.usdBalanceString
+                        : ''}
+                    </div>
+                  </div>
+                ))
```
[the remark this answers is not in the document]
Comment (Boyuan-Deng) [reply src/components/ZkAccoutModal/index.tsx]: done

File: src/components/ZkAccoutModal/index.tsx [hunk 29]
```diff
@@ -0,0 +1,192 @@
+import React, { useEffect, useState } from 'react';
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/ZkAccoutModal/index.tsx:1]: bump

File: src/components/ZkAccoutModal/index.tsx [hunk 30]
```diff
@@ -0,0 +1,192 @@
...
+  const assets = AssetType.AllCurrencies(true);
+  const [balanceFetching, setBalanceFetching] = useState(false);
+  const [balances, setBalances] = useState<any>([]);
+  const [totalBalanceString, setTotalBalanceString] = useState('$0.00');
+
+  const fetchPrivateBalances = async () => {
+    let total = new Decimal(0);
+    setBalanceFetching(true);
+    const balances = await Promise.all(
+      assets.map(async (assetType: any) => {
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/ZkAccoutModal/index.tsx:35]: bump

File: src/components/ZkAccoutModal/index.tsx [hunk 31]
```diff
@@ -0,0 +1,192 @@
...
+  const { usdPrices } = useUsdPrices();
+
+  const assets = AssetType.AllCurrencies(true);
+  const [balanceFetching, setBalanceFetching] = useState(false);
+  const [balances, setBalances] = useState<any>([]);
+  const [totalBalanceString, setTotalBalanceString] = useState('$0.00');
+
+  const fetchPrivateBalances = async () => {
+    let total = new Decimal(0);
+    setBalanceFetching(true);
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/ZkAccoutModal/index.tsx:33]: bump

File: src/components/ZkAccoutModal/index.tsx (hunk 20 above)
Comment (Kevingislason) [reply src/components/ZkAccoutModal/index.tsx]: Also, it would be nice to factor out the value returned from the map function to be its own component. In general, if you have a `map` inside JSX, and the component is longer than a few lines, you will want to factor out a subcomponent for improved readability.

File: src/components/ZkAccoutModal/index.tsx [hunk 32]
```diff
@@ -0,0 +1,175 @@
...
+                {balance.privateBalance.toString()}
+              </div>
+            </div>
+          </div>
+          <div className="text-fourth">
+            {usdPrices[balance.assetType.baseTicker]
+              ? balance.usdBalanceString
+              : '$0.00'}
+          </div>
+        </div>
```
Comment (Kevingislason) [inline src/components/ZkAccoutModal/index.tsx:157]: You should probably factor out the return of this map function into a subcomponent. In general, you will want to do this almost everywhere you have a `map` in JSX, for improved readability.

File: src/pages/SendPage/SendFromForm/PrivateFromAccountSelect.tsx [hunk 33]
```diff
@@ -1,46 +1,34 @@
...
-          />
-        )}
+    <>
+      <div className="px-6 py-5 leading-5 h-16 flex items-center gap-2 text-lg text-primary relative gradient-border rounded-full">
+        Private{' '}
+        <span
+          className="manta-gray text-xs cursor-pointer"
+          onClick={() => showAccountModal()}
+        >
+          {privateAddress?.slice(0, 10)}...{privateAddress?.slice(-10)}
```
Comment (Kevingislason) [inline src/pages/SendPage/SendFromForm/PrivateFromAccountSelect.tsx:24]: It would be nice to define a variable `privateAddresDisplayString` outside the JSX here. In general, it's best to leave all code logic outside of JSX for improved readability, even very simple logic like this.

**Review by Kevingislason — COMMENTED** [review]
Nice work. Code basically looks good, though I think you missed a few comments from earlier reviews, which I have bumped. 

Biggest outstanding issue: I think we had wanted to display a colored button and an informational modal when the signer is not connected. I'm pretty sure we had discussed this, do you recall? Perhaps you are not fully pushed?

File: src/components/ZkAccoutModal/index.tsx (hunk 31 above)
[the remark this answers is not in the document]
Comment (Boyuan-Deng) [reply src/components/ZkAccoutModal/index.tsx:33]: balanceFetching was not doing any meaning work so removed from the codebase

File: src/components/Modal/connectSigner.tsx [hunk 34]
```diff
@@ -0,0 +1,48 @@
...
+const ConnectSignerModal = () => {
+  return (
+    <div className="p-4 w-96 flex flex-col justify-center text-center text-sm">
+      <img
+        className="place-self-center w-12 h-12"
+        src={MantaIcon}
+        alt="Manta"
+      />
+      <p className="my-4 text-white">
+        Log in Manta Signer to see your zkAssets and transact it.
```
Comment (AlexParshh) [inline src/components/Modal/connectSigner.tsx:13]: This kind of sounds weird maybe replace it with something like: 
"Log in to Manta Signer to see your zkAssets and to start transacting."

File: src/components/Navbar/SignerConnectionStatusLabel.tsx [hunk 35]
```diff
@@ -62,6 +62,7 @@ SignerConnectedLabel.propTypes = {
 const SignerConnectionStatusLabel = () => {
   const config = useConfig();
   const { signerVersion } = usePrivateWallet();
+
```
Comment (Kevingislason) [inline src/components/Navbar/SignerConnectionStatusLabel.tsx:65]: We can delete this file now :)

File: src/contexts/privateWalletContext.tsx [hunk 36]
```diff
@@ -144,10 +149,13 @@ export const PrivateWalletContextProvider = (props) => {
...
-    const interval = setInterval(async () => {
-      fetchSignerVersion();
-    }, 2000);
-    return () => clearInterval(interval);
+    let interval;
+    // if (!signerIsConnected) {
+      interval = setInterval(async () => {
+        fetchSignerVersion();
+      }, 2000);
+    // }
```
Comment (Kevingislason) [inline src/contexts/privateWalletContext.tsx:157]: Remove comment

File: src/pages/SendPage/SendContext/index.tsx [hunk 37]
```diff
@@ -280,7 +280,7 @@ export const SendContextProvider = (props) => {
       fetchSenderBalance();
       fetchReceiverBalance();
       fetchFeeBalance();
-    }, 200);
+    }, 3000);
```
Comment (Kevingislason) [inline src/pages/SendPage/SendContext/index.tsx:283]: Just curious, why did you change this?

File: src/types/Usd.ts [hunk 38]
```diff
@@ -0,0 +1,35 @@
...
+    this.value = value;
+  }
+
+  add(val: Decimal) {
+    this.value = this.value.add(val);
+  }
+
+  minus(val: Decimal) {
+    this.value = this.value.minus(val);
+  }
```
Comment (Kevingislason) [inline src/types/Usd.ts:16]: these should take `Usd` rather than `Decimal` as their arguments

**Review by Kevingislason — COMMENTED** [review]
Nice work on this, just a few minor changes and we can merge.

- [ ] The signer install modal should have a little less padding to match the design
- [ ] See Alex's comment about changing some of the wording in the signer install modal
- [x] I would also change "try lower" to "try lowering" 
- [ ] We can remove the "How can I use Dolphin testnet" link
- [ ] The signer install modal and the zk account modal have different colors in the design
- [x] There should be a link icon next to "Why do I need Manta Signer"

Also one bug of notel:
- [ ] If I submit a public to private transaction, and then open the zk account modal, the modal shows no funds for a moment after the transaction succeeds

Design link for convenience:
https://www.figma.com/file/9nAYHsYlfHmyT8RL5P8ISZ/Manta-Pay

File: src/pages/SendPage/SendContext/index.tsx (hunk 37 above)
Comment (Boyuan-Deng) [reply src/pages/SendPage/SendContext/index.tsx:283]: I changed it to 1000 now. No specific reason but just feel unnecessary to fetch the state every 0.2 s but let me know if you feel that is necessary

File: src/types/Usd.ts [hunk 39]
```diff
@@ -7,20 +7,20 @@ export default class Usd {
...
-  mul(val: Decimal) {
-    this.value = this.value.mul(val);
+  mul(usd: Usd) {
+    this.value = this.value.mul(usd.value);
   }
 
-  div(val: Decimal) {
-    this.value = this.value.div(val);
+  div(usd: Usd) {
+    this.value = this.value.div(usd.value);
```
Comment (Kevingislason) [inline src/types/Usd.ts:23]: unlike add and minus, div and mul should take decimals rather than dollars as their argument; multiplying dollars x dollars doesn't make sense because you get dollars^2 units

File: src/components/Accounts/ZkAccountModal.tsx [hunk 40]
```diff
@@ -28,68 +28,63 @@ const AccountDisplay = () => {
...
-          assetType,
-          usdBalance,
-          usdBalanceString,
-          privateBalance
-        };
-      }
-
+  const fetchPrivateBalance = async (assetType: any) => {
+    let usdBalance = null;
+    console.log('1', assetType.baseName);
```
Comment (Kevingislason) [inline src/components/Accounts/ZkAccountModal.tsx:33]: please delete the logs in this file

File: src/components/Accounts/ZkAccountModal.tsx [hunk 41]
```diff
@@ -28,68 +28,63 @@ const AccountDisplay = () => {
...
+    const privateBalance = await getSpendableBalance(assetType);
+    console.log('2', assetType.baseName);
+    if (privateBalance) {
+      console.log('3', assetType.baseName);
+      const assetUsdValue = usdPrices[assetType.baseTicker] || new Decimal(0);
```
Comment (Kevingislason) [inline src/components/Accounts/ZkAccountModal.tsx:38]: shouldn't `new Decimal(0)` be a `Usd`?

File: src/components/Accounts/ZkAccountModal.tsx [hunk 42]
```diff
@@ -28,68 +28,63 @@ const AccountDisplay = () => {
...
   };
 
   const fetchPrivateBalances = async () => {
     let totalUsd = new Usd(new Decimal(0));
+    const nonzeroBalance = (balance) =>
+      balance.privateBalance &&
+      balance.privateBalance.gt(new Balance(balance.assetType, new BN(0)));
+    console.log(`
+      assets - ${assets.length}
+    `);
```
Comment (Kevingislason) [inline src/components/Accounts/ZkAccountModal.tsx:64]: I think you misunderstood my comment, I wasn't suggesting you make a nonzeroBalance function, but just that you do something like this to improve readability:

```js
const nonzeroBalances =  [
      ...updatedBalances.filter(
        (balance) =>
          balance.privateBalance &&
          balance.privateBalance.gt(new Balance(balance.assetType, new BN(0)))
      )
      ...updatedBalances.filter((balance) => nonzeroBalance(balance))
];
setBalances(nonzeroBalances)
```

**Review by Kevingislason — COMMENTED** [review]
Good stuff, comments + a few UI tweaks and this is set
- [x] There are two periods in "Log in to Manta Signer to see your zkAssets and to start transacting.."
- [x] Let's shorted the Brave text to "For Brave users: if you are already logged in, lower shields to connect"
- [x] I think it would look better if the "Sign In" text is a bit bigger
- [x] I think the component looks a bit too wide, let's add back a little vertical padding

File: src/contexts/privateWalletContext.tsx [hunk 43]
```diff
@@ -189,6 +197,10 @@ export const PrivateWalletContextProvider = (props) => {
   }, [isReady]);
 
   const getSpendableBalance = async (assetType) => {
+    console.log(`
+    isReady - ${isReady}
```
Comment (AlexParshh) [inline src/contexts/privateWalletContext.tsx]: remove logging

**Review by AlexParshh — COMMENTED** [review]
looks good overall, I agree with Kevin's requested changes above

**Review by AlexParshh — APPROVED** [review]

File: src/types/Usd.ts (hunk 39 above)
Comment (Boyuan-Deng) [reply src/types/Usd.ts]: lol, funny math mistake

File: src/components/Accounts/ZkAccountModal.tsx [hunk 44]
```diff
@@ -0,0 +1,100 @@
...
+const PrivateTokenTableContent = ({ balances }) => {
+  const { txStatus } = useTxStatus();
+  const privateWallet = usePrivateWallet();
+  if (balances && balances.length > 0) {
+    return balances.map((balance, index) => (
+      <PrivateTokenItem balance={balance} key={balance.assetType.assetId} />
+    ));
+  } else if (
+    privateWallet?.balancesAreStale?.current ||
+    txStatus?.isProcessing()
```
Comment (Kevingislason) [inline src/components/Accounts/ZkAccountModal.tsx]: this is no longer necessary, we now show balances even when a tx is processing

**Review by Kevingislason — APPROVED** [review]
did a few last minor cleanup things myself
-tweaked signer out of date text
-imported React in some components where it was missing
-removed some unused imports
-removed one condition where we tried to display loading state in the zk account modal while a transaction is processing. this is no longer necessary
